### DISCUSSION TITLE
Custom data types and optional annotations

### DISCUSSION THREAD

**Discussion by quantumtrope** [pydantic/pydantic#4929] [Question] [answered]
I use [`pint`](https://pint.readthedocs.io/en/stable/index.html) to handle units in a lot of my work. I like using `pydantic` to generate models of what I expect. `pint` has a functionality to check whether a given `Quantity` matches a [specific dimensionality](https://pint.readthedocs.io/en/stable/advanced/wrapping.html#checking-dimensionality). 

The question is: how could I provide some form of metadata do the pydantic custom data type to constrain to a specific dimensionality like `[length]`? For example, if I have a class `PintQuantity` that properly implements a `__get_validators__` and `validate` function as per [the docs](https://docs.pydantic.dev/usage/types/#custom-data-types), how could I write something like:

```python
class TestClass(BaseModel):
    q: PintQuantity["length"]
```
to indicate that I want to only accept quantities which are somehow a length: "meter","foot","inch" etc. I know how I would change the validator on the `PintQuantity` class. This question is about how to provide that additional constraint information to my custom data class. 

Is this possible in `pydantic` V1 or do I need to wait till V2?

**Comment by uriyyo** [comment, marked as answer]
Hi @quantumtrope,

As an option you can override `__class_getitem__` method in `PintQuantity` class:
```py
from typing import ClassVar, Any

from pydantic import BaseModel


class PintQuantity:
    __dimensionality__: ClassVar[str | None] = None

    def __class_getitem__(cls, item):
        return type(cls.__name__, (cls,), {"__dimensionality__": item})

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v: Any) -> Any:
        # your validation logic here
        print(f"Validating {v} as {cls.__dimensionality__}")
        return v


class TestClass(BaseModel):
    q: PintQuantity["length"]
    v: PintQuantity["velocity"]


TestClass(q=1, v=2)
```
Code above will produce output like this:
```py
Validating 1 as length
Validating 2 as velocity
```

**Reply by quantumtrope** [reply]
@uriyyo -- thanks for the suggestion! Works like a charm. I learned something new too about how to create a generic class for typing. Very helpful!

**Comment by MichaelTiemannOSC** [comment]
Anybody taken a crack at answer this question for Pydantic 2.x?

**Reply by samuelcolvin** [reply]
Should be the same except the validators definition use V2 style, we now have support for v1 custom types in V2, so it might just work.

**Comment by MichaelTiemannOSC** [comment]
xref https://github.com/hgrecco/pint/issues/1770#issuecomment-1713892219

@samuelcolvin I'm sure the above goes against the Pydantic grain six ways from Sunday.  Appreciate any comments you'd like to make about how to make it go better with the grain.
